Pass on the new screen-reader section (#634)
browserengineering/book: book/accessibility.md

The user can now direct the screen reader to walk up or down this
accessibility tree and describe each node to the user. 

Screen readers
==============

Typically, the screen reader is a separate application from the

> browser,[^why-diff] which the browser communicates with through

Reviewer: ```suggestion
browser,[^why-diff] with which the browser communicates through
```

```

Now, what should the screen reader say? Well, that's not really up to
the browser---the screen reader is a stand-alone application, often
heavily configured by its user, and can decide on its own. But as a
simple debugging aid, let's write a screen reader that speaks the
whole web page once it's loaded; of course, a real screen reader is
much more flexible than that.

To speak the whole document, we need to know how to speak each
`AccessibilityNode`. This has to be decided back in the `Tab`, since
the text will include DOM content that is not accessible to the
browser thread. So let's add a `text` field to `AccessibilityNode` and
set it according to role and surrounding DOM context. For each node,

> we'll figure out its text via the `announce_text` function. For text

Reviewer: Rephrase to remove `announce_text`

            new_text = accessibility_node.text
            if new_text:
                text += "\n"  + new_text

        speak_text(text)

    def speak_update(self):
        if not self.accessibility_tree: return

        if not self.has_spoken_document:
            self.speak_document()
            self.has_spoken_document = True

```

Speaking the whole document happens only once. But the user might need

> feedback as they browse the page. For example, let's speak the focused

Reviewer: ```suggestion
feedback as they browse the page, such as which element is focused. Let's speak the focused
```

Author: I rephrased in a different way.

field on `Browser` with the last focused element we actually spoke out
loud:

``` {.python}
class Browser:
    def __init__(self):
        # ...
        self.last_tab_focus = None
```

Then, if `tab_focus` isn't equal to `last_tab_focus`, we know focus
has moved and it's time to speak the focused node:

``` {.python}
class Browser:
    def speak_update(self):
        if self.tab_focus and \
            self.tab_focus != self.last_tab_focus:
            nodes = [node for node in tree_to_list(self.accessibility_tree, [])
                        if node.node == self.tab_focus]
            if nodes:
                self.speak_node(nodes[0], "element focused ")
            self.last_tab_focus = self.tab_focus
```

The `speak_node` method is similar to `speak_document` but it only
speaks a single node:

``` {.python}
class Browser:
    def speak_node(self, node, text):
        text += node.text
        if text and node.children and \
            node.children[0].role == "StaticText":
            text += " " + \
            node.children[0].text

        if text:
            speak_text(text)
```

There's a lot more in a real screen reader: landmarks, navigating text

> at different granularities, repeating recently spoken text, and so on.

Reviewer: *not* repeating recently spoken text?

Author: Screen readers have key bindings for "repeat that word again" or "repeat that sentence you just said", which is what I meant to point to here. Can you suggest a better phrasing?

loud:

``` {.python}
class Browser:
    def __init__(self):
        # ...
        self.last_tab_focus = None
```

Then, if `tab_focus` isn't equal to `last_tab_focus`, we know focus
has moved and it's time to speak the focused node:

``` {.python}
class Browser:
    def speak_update(self):
        if self.tab_focus and \
            self.tab_focus != self.last_tab_focus:
            nodes = [node for node in tree_to_list(self.accessibility_tree, [])
                        if node.node == self.tab_focus]
            if nodes:
                self.speak_node(nodes[0], "element focused ")
            self.last_tab_focus = self.tab_focus
```

The `speak_node` method is similar to `speak_document` but it only
speaks a single node:

``` {.python}
class Browser:
    def speak_node(self, node, text):
        text += node.text
        if text and node.children and \
            node.children[0].role == "StaticText":
            text += " " + \
            node.children[0].text

        if text:
            speak_text(text)
```

There's a lot more in a real screen reader: landmarks, navigating text
at different granularities, repeating recently spoken text, and so on.

> Those features make various use of the accessibility tree and the

Reviewer: ```suggestion
Those features make various uses of the accessibility tree and the
```
